web/js/codeworld.js

```diff
@@ -13,17 +13,148 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-'use strict';
+
+import {
+  createFolder,
+  definePanelExtension,
+  deleteFolder_,
+  deleteProject_,
+  discoverProjects,
+  getNearestDirectory,
+  initDirectoryTree,
+  loadProject,
+  loadSample,
+  markFailed,
+  onHover,
+  parseCompileErrors,
+  parseSymbolsFromCurrentCode,
+  printMessage,
+  registerStandardHints,
+  renderDeclaration,
+  run,
+  saveProjectBase,
+  saveProjectAsBase,
+  setCode,
+  share,
+  shareFolder_,
+  toggleObsoleteCodeAlert,
+  updateTreeOnNewProjectCreation,
+  warnIfUnsaved,
+} from './codeworld_shared.js';
+import * as TreeDialog from './treedialog.js';
+
+import * as Alert from './utils/alert.js';
+import * as Auth from './utils/auth.js';
+import * as DirTree from './utils/directoryTree.js';
+import { sendHttp } from './utils/network.js';
+import { onObjectPropertyChange } from './utils/object.js';
+
+init();
+
+function attachEventListeners() {
+  $('#signout').on('click', () => {
+    Auth.signOut(isEditorClean, clearWorkspace, () => {
+      $('#nav').trigger('disable');
+    });
+  });
+  $('#signin').on('click', () => {
+    Auth.signIn(() => {
+      discoverProjects('');
+
+      $('#nav').trigger('enable');
+    });
+  });
+
+  $('#newButton').on('click', newProject);
+  $('#newFolderButton').on('click', newFolder);
+  $('#deleteButton').on('click', deleteProject);
+  $('#saveButton').on('click', saveProject);
+  $('#saveAsButton').on('click', saveProjectAs);
+  $('#downloadButton').on('click', downloadProject);
+  $('#docButton').on('click', help);
+  $('#toggleThemeButton').on('click', toggleTheme);
+  $('#shareFolderButton').on('click', shareFolder);
+
+  $('#startRecButton').on('click', captureStart);
+  $('#stopRecButton').on('click', stopRecording);
+  $('#shareButton').on('click', share);
+  $('#inspectButton').on('click', inspect);
+
+  $('#runButton').on('click', compile);
+  $('#stopButton').on('click', stopRun);
+
+  $('#runner').on('load', toggleObsoleteCodeAlert);
+}
+
+function attachCustomEventListeners() {
+  $('#nav').on('enable', () => {
```

Comment:
I think this is what I've been trying to avoid, actually.  I think it's more important to get this merged than to get everything right, so I'm not insisting on changes right now.  But I think, like in the past, I really want one source of truth for everything, and the source of truth for whether you're signed in should be `Auth.signedIn()`.

So instead of saying "when nav is enabled, do these things", we should really say "when you're signed in, do these things".  The logic should be idempotent, so calling it twice is fine.  Then you could add a listener that gets called whenever the signed-in state changes, that runs that code.

The `onObjectPropertyChange` sections below are perfect, and get this right.  The underlying state changes should drive UI changes.  UI changes shouldn't trigger other UI changes.

But, not a big deal to do all this right away.

Reply:
I probably misunderstood your intention to implement custom events. `$('#nav').trigger('event_name')` simply dispatches an event, there are implied UI actions when calling this method.

Actually, we already define a listener that gets called whenever user's logged in state changes - I moved respective actions there.